server/src/controller/product.controller.ts

```diff
@@ -0,0 +1,186 @@
+            if (image !== '') await deleteFile(image);
+            return next(new ErrorHandler(errorStatusCode.BadRequest, errorMsg.ParameterError));
+        }
+        try {
+            
+            const checkforeignKeyExit = await this.storeService.getById(storeId);
+            if (!checkforeignKeyExit) {
+                return next(new ErrorHandler(errorStatusCode.BadRequest, errorMsg.StoreIdError));
+            }
+            /** 確認 name 是否有重複命名 */
+            const checkExistByName = await this.service.checkExistByName(name,storeId);
+            if(checkExistByName){
+                return next(new ErrorHandler(errorStatusCode.BadRequest, errorMsg.DataAlreadyExist));
+            }
+            let tagsData: Tag[] | undefined;
+            if (tags) {
+                tagsData = await this.tagService.getByIds(<number[]>tags);
+                if (!tagsData) {
+                    return next(new ErrorHandler(errorStatusCode.BadRequest, errorMsg.TagAssociationError));
+                }
+            }
+            tags = tagsData;
+            const params = { name, money, image, storeId, category, status, tags };
+            const newProduct = await this.service.create(params);
+            res.status(200).json({ result: true });
+        } catch (error) {
+            console.log("create db error: ", error);
+            return next(new ErrorHandler(errorStatusCode.InternalServerError, errorMsg.InternalServerError));
+        }
+    }
+
+    async update(req: Request, res: Response, next: NextFunction) {
+        const id: number = parseInt(req.params.id);
+        if (!id) {
+            return next(new ErrorHandler(errorStatusCode.InternalServerError, errorMsg.InternalServerError));
+        }
+        const image = req.file ? req.file.filename : '';
+        let { name, storeId, money, category, status, tags }: IProductUpdateParams = req.body;
+        if (!name || !money || status === undefined || status === undefined) {
+            if (image !== '') await deleteFile(image);
+            return next(new ErrorHandler(errorStatusCode.BadRequest, errorMsg.ParameterError));
+        }
+        try {
+            const checkforeignKeyExit = await this.storeService.getById(storeId);
+            if (!checkforeignKeyExit) {
+                return next(new ErrorHandler(errorStatusCode.BadRequest, errorMsg.StoreIdError));
+            }
+
+            let tagsData: Tag[] | undefined;
+            if (tags) {
+                tagsData = await this.tagService.getByIds(<number[]>tags);
+                if (!tagsData) {
+                    return next(new ErrorHandler(errorStatusCode.BadRequest, errorMsg.TagAssociationError));
+                }
+            }
+            const product = await this.service.getById(id);
+            if (!product) {
+                return next(new ErrorHandler(errorStatusCode.BadRequest, errorMsg.DataNotFound));
+            }
+            const newTagId: number[] = <number[]>tags
```

Comment:
this block looks weird. product is optional, so if product == undefined or null something, it will causes error ?